compiler/rustc_codegen_llvm/src/intrinsic.rs

```diff
@@ -1839,6 +1839,16 @@ unsupported {} from `{}` with element `{}` of size `{}` to `{}`"#,
         simd_neg: Int => neg, Float => fneg;
     }
 
+    if name == sym::simd_arith_offset {
+        let pointee = in_elem
+            .builtin_deref(true)
+            .unwrap_or_else(|| span_bug!(span, "must be called with a vector of pointer types"));
+        let layout = bx.layout_of(pointee.ty);
+        let ptrs = args[0].immediate();
+        let offsets = args[1].immediate();
+        return Ok(bx.gep(bx.backend_type(layout), ptrs, &[offsets]));
```

Comment:
Does `gep` type check this? Or can calling the intrinsic improperly result in an LLVM codegen error?

Reply:
I don't know, so we should probably add a codegen test rather than finding out the hard way. `gep` isn't that complex but it's known to ambush unwary compiler engineers when they turn their back on it.

Reply:
I added a check for the integer type in the second operand.

Reply:
Oh, I mostly just want to see with this that we generate a vector of pointers to the given type and then gep it like we Damn Well Should and don't suddenly somehow revert to scalar operations or something like that. A smoke test that the suite of intrinsics used to do a gather or scatter compiles correctly, basically, and that the types don't go suddenly weird on us.

> This does about as much checking as the arith_offset intrinsic, I think. The one extra check there is that the offsets must be ptr-sized integers, which arith_offset enforces via the type signature (but I am not sure if anything checks that type signature). I can try to add that here.

For the `#[repr(simd)]` types, while we do type-checking in rustc's front and "middle" phases to guarantee the input vector types are valid machine vector types, you can assume that it probably has bypassed any sensible checking like "is this even actually a pointer?" and is relying heavily on correct usage. Thus we would like to error during monomorphization on anything fishy.

Reply:
And if that doesn't quite land as coherent (which is fine):

I do not believe we have anything that would enforce that the first arg to simd_arith_offset would be a vector of pointers.

Reply:
> Oh, I mostly just want to see with this that we generate a vector of pointers to the given type

Okay... I'll try my best but I barely ever work with codegen tests so I am not even sure of the syntax to use.^^
Should it go through portable-simd APIs or call the intrinsics directly?

> A smoke test that the suite of intrinsics used to do a gather or scatter compiles correctly, basically, and that the types don't go suddenly weird on us.

We have the doctests as smoke tests as well. ;)